pkg/kubelet/images/pullmanager/image_pull_manager.go

```diff
@@ -0,0 +1,472 @@
+
+	imageService kubecontainer.ImageService
+
+	intentAccessors *StripedLockSet // image -> sync.Mutex
+	intentCounters  map[string]int  // image -> number of current in-flight pulls
+
+	pulledAccessors *StripedLockSet // imageRef -> sync.Mutex
+}
+
+func NewImagePullManager(ctx context.Context, recordsAccessor PullRecordsAccessor, imagePullPolicy ImagePullPolicyEnforcer, imageService kubecontainer.ImageService) (*PullManager, error) {
+	m := &PullManager{
+		recordsAccessor: recordsAccessor,
+
+		requireCredentialVerification: imagePullPolicy,
+
+		imageService: imageService,
+
+		intentAccessors: NewStripedLockSet(defaultLockSetSize),
+		intentCounters:  make(map[string]int),
+
+		pulledAccessors: NewStripedLockSet(defaultLockSetSize),
+	}
+
+	m.initialize(ctx)
+
+	return m, nil
+}
+
+func (f *PullManager) RecordPullIntent(image string) error {
+	f.intentAccessors.Lock(image)
+	defer f.intentAccessors.Unlock(image)
+
+	if err := f.recordsAccessor.WriteImagePullIntent(image); err != nil {
+		return fmt.Errorf("failed to record image pull intent: %w", err)
+	}
+
+	f.intentCounters[image]++
+	return nil
+}
+
+func (f *PullManager) RecordImagePulled(image, imageRef string, credentials *kubeletconfiginternal.ImagePullCredentials) {
+	if err := f.writePulledRecord(image, imageRef, credentials); err != nil {
+		klog.ErrorS(err, "failed to write image pulled record", "imageRef", imageRef)
+		return
+	}
+
+	f.decrementImagePullIntent(image)
+}
+
+// writePulledRecord writes an ImagePulledRecord into the f.pulledDir directory.
+// `image` is an image from a container of a Pod object.
+// `imageRef` is a reference to the `image“ as used by the CRI.
+// `credentials` is a set of credentials that should be written to a new/merged into
+// an existing record.
+//
+// If `credentials` is nil, it marks a situation where an image was pulled under
+// unknown circumstances. We should record the image as tracked but no credentials
+// should be written in order to force credential verification when the image is
+// accessed the next time.
+func (f *PullManager) writePulledRecord(image, imageRef string, credentials *kubeletconfiginternal.ImagePullCredentials) error {
+	f.pulledAccessors.Lock(imageRef)
+	defer f.pulledAccessors.Unlock(imageRef)
+
+	sanitizedImage, err := trimImageTagDigest(image)
+	if err != nil {
+		return fmt.Errorf("invalid image name %q: %w", image, err)
+	}
+
+	pulledRecord, _, err := f.recordsAccessor.GetImagePulledRecord(imageRef)
+	if err != nil {
+		klog.InfoS("failed to retrieve an ImagePulledRecord", "image", image, "err", err)
+		pulledRecord = nil
+	}
+
+	var pulledRecordChanged bool
+	if pulledRecord == nil {
+		pulledRecordChanged = true
+		pulledRecord = &kubeletconfiginternal.ImagePulledRecord{
+			LastUpdatedTime:   metav1.Time{Time: time.Now()},
+			ImageRef:          imageRef,
+			CredentialMapping: make(map[string]kubeletconfiginternal.ImagePullCredentials),
+		}
+		if credentials != nil {
```

Comment:
comment that the existence of the pulled record for the imageRef is sufficient to indicate this was a kubelet-pulled image, and if there's no entry for the image in the credential mapping, the kubelet fails safe and requires credential validation